src/platform/posix/tcp_ip_channel.c

```diff
@@ -308,12 +284,12 @@ static lf_ret_t TcpIpChannel_send_blocking(NetworkChannel *untyped_self, const F
   int timeout = TCP_IP_CHANNEL_NUM_RETRIES;
 
   while (bytes_written < message_size && timeout > 0) {
-    LF_DEBUG(NET, "Sending %d bytes", message_size - bytes_written);
+    LF_DEBUG(NET, "TcpIpChannel: Sending %d bytes", message_size - bytes_written);
     ssize_t bytes_send = send(socket, self->write_buffer + bytes_written, message_size - bytes_written, 0);
-    LF_DEBUG(NET, "%d bytes sent", bytes_send);
+    LF_DEBUG(NET, "TcpIpChannel: %d bytes sent", bytes_send);
 
     if (bytes_send < 0) {
-      LF_ERR(NET, "write failed errno=%d", errno);
+      LF_ERR(NET, "TcpIpChannel: Write failed errno=%d", errno);
       switch (errno) {
       case ETIMEDOUT:
       case ENOTCONN:
```

Comment:
Is it necessary to call `_TcpIpChannel_update_state` here? We might get some very subtle races here when the worker thread also detects that the socket is closed he will initiate a reconnect right? Then we don't want to overwrite the new state if we fail to send some data as well?

If we are guaranteed that the worker thread will not block forever on `receive`, and that it will detect anything that is detected here, then I propose that we only read the channel state from the runtime side, and only write to it from the worker thread. What do you think?

Reply:
Yes there might be a race if the worker thread already started a reconnection. We could guard this though.
I think it is better to update the state straight away if we see something is wrong.

A very robust way could be to just mutex the whole worker thread `while loop` and then also mutex the whole send_blocking function? And of course also the `get_state` function. This way both should not be able to interfer with each other.
We should though then change back to making the tcp `receive` non-blocking.

Reply:
I have now implemented a possible solution in the [Fix race conditions](https://github.com/lf-lang/reactor-uc/pull/150/commits/373782dec4f7a724053d82a2815f691ded470876) commit